### ISSUE TITLE
Pattern for not passing `continue` callbacks

### ISSUE THREAD

**Issue by louisameline** [vue-sync/rfcs#20]
Currently:
```javascript
function storeUpdateFn (_doc) {
  ...
}

if (isFunction(syncHookFn)) {
  syncHookFn(storeUpdateFn, doc, id, store)
} else {
  storeUpdateFn(doc)
}
````
and on the user side:
```javascript
addedHook: function (updateStore, doc, id, store) {
  ...
  return updateStore(doc)
}
```
Tomorrow:
```javascript
function storeUpdateFn (_doc) {
  ...
}

const hook = syncHookFn
  ? syncHookFn.bind(null, doc, id, store)
  : true

Promise.resolve()
  .then(hook)
  .then(_doc => storeUpdateFn(_doc || doc))
  // the fact that we put a catch here or not will depend on how we deal
  // with error handling
  .catch(err => {})
````
and on the user side:
```javascript
addedHook: function (doc, id, store) {
    ...
    // true to continue, false to stop, or a promise which resolves
    // with the doc to proceed with, or which rejects
    return true | false | Promise
}
```
I found out yesterday the `return` that we need to put in front of `updateStore(doc)` makes it unfeasible to allow the hook to respond asynchronously, so my PR would not work, so that's another reason not to use it. The above syntax with a promise is required to make it work.
But since the new signature is a breaking change, I won't make a PR for v1.

**Comment by mesqueeb** [comment]
I see!

With the current 2.0 syntax I expose an `abort` function that can be executed.

I believe "returning" inside an event felt farfetched.

The way I expose `abort` is eg. like so:

```js
collection('pokedex').insert({ name: 'Louis' }, {
  on: { before: async ({ abort }) => {  } }
})
```

Here is a quick overview of the events and params:

- `before`
- `success`
- `error`

All these can be passed a normal or async function.
That means, they can return either `void` or `Promise<void>`.

All of the event functions will get these params in an object:

- `payload` the payload of the action
- `actionName` the actionName that was triggered (get, insert, ...)
- `storeName` the plugin store name (eg. "local" or "vuex", depending on the name chosen by the dev)
- `abort` a function that can be triggered to abort the action chain.

When `abort` is triggered after eg. the local store is updated, but before the remote store is triggered, core will stop execution and not talk to the remote store.

If you want to `abort` but also throw an error, you can simply throw an error inside the event, you don't need to trigger `abort` at all.

I feel that presenting the dev with an `abort` function that can be triggered is much more clear than returning a boolean.

### error handling

If there is an error in any of the stores you can choose to "stop" execution, "revert" execution or "continue" execution.

Let me explain by example:

```js
collection('pokedex').insert({ name: 'Louis' })
  .then(() => console.log('ok'))
  .catch((error) => console.error('oh no!', error))
```

If you have 1 local store, 1 remote store, and the write order is `["local", "remote"]`, and there is an error during the "remote" store.

What happens by default is that "local" will have already added the data, but, because remote has an error, "core" lets the local store plugin know he needs to revert. The local store plugin reverts the data based on what information he has (the action that failed, the payload, etc.).

Then that error will still propagate and be thrown, so you can "catch" it even thought the store was reverted for you.

In case you want `onError: 'stop'` the local store will not be reverted, and the error will propagate so you can catch it.

In case you want `onError: 'continue'`, nothing will happen, local store stays as it was, core continues on to execute the next store if there is any, if not the action resolves.

If you choose `onError: 'continue'` however, you can still catch the error inside a "error" event.

### hooks on writing data

I thought it was weird having events be responsible for returning data.

For modifying the payload a separate function can be provided eg.:

```js
const pokedex = vueSync.collection('pokedex', {
  modifyPayloadOn: {
    // apply default values on insert
    insert: (payload) => { return { archived: false, ...payload } },
    // any action name possible
  }
})
```
in this case, the modify payload function will be triggered once on the payload before it is sent to any store plugin.

### hooks on reading data

For modifying whatever data gets "read" from the server, before it's inserted locally, I prepared this config:

```js
const pokedex = vueSync.collection('pokedex', {
  modifyReadResponseOn: {
    // apply default values when reading docs from the server
    added: (payload) => { return { archived: false, ...payload } },

    // also possible: 'modified' and 'removed'
  }
})
```
In this case, 'added' gets triggered on `get()` actions. And any of the 3 types can get triggered during streams.

**Comment by louisameline** [comment]
I need to think more about this. Events have always been easier to work with, and having two systems in parallel with the modify payload thing feels awkward.

It's ok for events to bubble up and down the chain of listeners and be able to return something. Think about DOM event listeners:
- you can create and modify properties on the `event` object, information which will be thus available to next listeners
- the listener may return false to prevent further bubbling
- you may call methods on the event like `event.stopPropagation`, `event.stopImmediatePropagation`, `event.preventDefault`, etc.
- the event bubbles up AND down (capturing) the chain of listeners (might be useful to us, to be seen)
- the error itself should be transmitted as an event
- listeners can be bound and unbound at any time from any component of the chain, very flexible

Events feel natural to me.

I also remember explaining in an issue how relying on promises (it was in openDBChannel) sometimes didn't work because of race conditions. The event system felt compelling there.

Also, the event handlers could be created in the app or in vuex action files, and did not required to be listed in the module definition file, that was important. But maybe they can also be with your syntax, I haven't checked.

I reserve my judgment yet, but I feel there is a strong case for events-only insteads of two systems in parallel.

**Comment by mesqueeb** [comment]
I'm not 100% sure what kind of syntax you were looking for when you want something "closer to events".

Can you give an example?

I also wanna give some more thoughts on these:

> I also remember explaining in an issue how relying on promises (it was in openDBChannel) sometimes didn't work because of race conditions. The event system felt compelling there.

I'm not sure what you mean with this as well.

But, with the current implementation, as I explained, you can make it so your event handler returns a promise. In this case Vue Sync will wait for that event promise to resolve before going on. There's no race conditions involved in the current implementation.

> Also, the event handlers could be created in the app or in vuex action files, and did not required to be listed in the module definition file, that was important. But maybe they can also be with your syntax, I haven't checked.

Yeah, the way I'm picturing the Vuex Plugin is that we use [Vuex Dynamic Module Registration](https://vuex.vuejs.org/guide/modules.html#dynamic-module-registration).

We basically give devs a function that will register the Vuex modules for the dev and also instantiate the VueSync module (collection or doc) under the hood. So this is one of the reasons you have complete control over how you want to expose the VueSync syntax. So you can have the users register events in whatever way you want them to (eg. inside a vuex module) and then have the plugin pass those events on to the VueSync collection underneath.

It basically would look like this:

```js
const myTraditionalVuexModule = { actions, state, mutations, getters }
vuexPlugin.registerModule(myTraditionalVuexModule)
```

Now, this being said. Can you give an example of the way you'd want the dev to be able to register an event handler?

**Comment by louisameline** [comment]
The race condition I am referring to was detailed in this issue: https://github.com/vue-sync/rfcs/issues/14#issuecomment-568302539

But now that I remember it well, I realize that's it's rather a Vuex issue (we can't trust the `delete` action promise to be resolved soon enough, because of some `$nextTick` somewhere in Vuex). I've solved it in my own v-e-f fork for the time being with an ugly patch involving a callback, so I never made a PR. I was planning to solve it elegantly with events, but I admit that it's something that concerns the user API of VueSync Vuex plugin more than the communication with the core itself. If we can find a pattern that applies everywhere though, it will be nice :)

Just for the record and to easily find old messages, events were also discussed here https://github.com/vue-sync/rfcs/issues/8#issuecomment-571591028

To come back to the event subject, from what I understood earlier, I was under the impression that you had a double event listener system. Sorry if I got it wrong!
One seemed to truly act as an event system, with the event being passed from one listener to the other, seemingly a "read-only" system, and the other which looked rather like the hook system in disguise to update the payload ;)

Intuitively, I expected a unified way of communication, something like
```javascript
// note: as discussed previously, this syntax is nice at module declaration, but
// we might need to also provide an emitter object to allow the classic
// `module.on(eventType, fn)` event-driven syntax for later binding, we'll see
on: {
  before (event) {
     if (payloadNeedsSynchronousModif) {
       // edit the payload and pass on
       event.payload = edit(event.payload)
       return true
    }
    else if (payloadNeedsAsyncModif) {
      const promise = doSomethingAsynchronous(event.payload)
      // the promise must resolve with the value of the modified (or not
      // modified) payload
      return promise
    }
    else if (mustStop) {
        // the way this works will require a discussion on its own. It might
        // trigger a new event which will be handled by all listeners that received
        // the current event before it got reverted, in reverse order?
        // I have also another idea, to let the event cycle continue with notification
        // of failure, we'll discuss it in a separate issue
        event.revert()
        // prevents the event from bubbling any further (= aborts)
        return false
      }
   }
}
```
But again, there are so many things to think of. Rollbacks, transactions, transmission of events from a Collection module `streaming` to its registered Doc modules for processing, collection group queries, and on and on... There is no way we can settle for good before we have started implementing the whole shebang and seen how it performs :)
Gosh I'm excited about this. A challenge for the mind :D
